sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,455 @@
+        to notify the server that negotiation is complete.
+
+        The capability negotiation is complete when all capability requests
+        have been either acknowledged or denied successfuly (directly or by
+        calling the :meth:`resume` method).
+        """
+        return all(
+            status
+            for cap_req, values in self._registered.items()
+            for _plugin_capability, status in values.values()
+            if cap_req in self._requested
+        )
+
+    # tell if registered, requested, acknowledged, or denied
+
+    def is_registered(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is registered.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._registered
+
+    def is_requested(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is requested.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._requested
+
+    def is_acknowledged(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is acknowledged.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._acknowledged
+
+    def is_denied(self, request: Iterable[str]) -> bool:
+        """Tell if a capability request is denied.
+
+        :param request: a set of capabilities that form a capability request
+                        together; this can be any sortable iterable
+        """
+        return tuple(sorted(request)) in self._denied
+
+    # register, request, resume, acknowledge, deny, etc.
+
+    def register(self, plugin_name: str, request: capability) -> None:
+        """Register a capability ``request`` for ``plugin_name``.
+
+        :param request: the capability request to register for later
+
+        Once registered, the capability request can be requested by the bot. A
+        registered request appears in :attr:`registered`::
+
+            >>> from sopel import plugin
+            >>> request = plugin.capability('cap1')
+            >>> manager.register('coretasks', request)
+            >>> ('cap1',) in manager.registered
+            True
+            >>> manager.is_registered(('cap1',))
+            True
+
+        It is not, however, directly requested::
+
+            >>> manager.is_requested(('cap1',))
+            False
+
+        See :meth:`request_available` to automatically request capabilities
+        advertised by the server.
+        """
+        self._registered.setdefault(request.cap_req, {})[plugin_name] = (
+            request, False,
+        )
```

Comment:
Nitpick: I would prefer this to be two statements with an additional name

```suggestion
        plugin_caps = self._registered.setdefault(request.cap_req, {})
        plugin_caps[plugin_name] = (
            request, False,
        )
```